### ISSUE TITLE
Make put() and post() more similar

### ISSUE THREAD

**Issue by Michael-Hamburger** [deployd/deployd#769]
At first I thought about asking this on gitter but after writing everything out it is longer than I thought. So here it is:

```Put``` and ```Post``` are similar. ```Post``` creates a new entry in the collection, ```Put``` updates an already existing entry in the collection. In fact ```Put``` only needs to load the already existing entry and just merges the changes. A coworker mentioned to me that the way deployd handles ```Put``` and ```Post``` is different and I looked at how they are are implemented. They are both methods in https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js.

```Put``` queries the entry, merges the changes, validates the collection (on validate event) and after the validation doesn't throw any error runPutEvent() is called.
runPutEvent() is then calling the on put event and when everything is correct commit() is called.

```Post``` calls on post event and when this completes without any error commit() is called.

As you can see ```Post``` never calls on validate event.
In fact the on validate event is called in https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L638 when the script exists.

Wouldn't it be cleaner to call the on validate event also in the post() method (like here in put() https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L582)  and remove it here https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L638 ?

The second conditional branch (https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L635) could be removed (because validate is called in post()). 

 ```
 if (query.id) {
    beforeRequestDomain.event = "PUT";
    collection.doBeforeRequestEvent(ctx, beforeRequestDomain, function(err) {
      if (err) return fn(err);
      put();
    });
  } else {
    collection.doBeforeRequestEvent(ctx, beforeRequestDomain, function(err) {
      if (err) return fn(err);
      post();
    });
  }
```

instead of

```
if (query.id) {
    beforeRequestDomain.event = "PUT";
    collection.doBeforeRequestEvent(ctx, beforeRequestDomain, function(err) {
      if (err) return fn(err);
      put();
    });
  } else if (collection.shouldRunEvent(collection.events.Validate, ctx)) {
    collection.doBeforeRequestEvent(ctx, beforeRequestDomain, function(err) {
      if (err) return fn(err);
      collection.events.Validate.run(ctx, domain, function (err) {
        if(err || domain.hasErrors()) return done(err || errors);
        post();
      });
    });
  } else {
    collection.doBeforeRequestEvent(ctx, beforeRequestDomain, function(err) {
      if (err) return fn(err);
      post();
    });
  }
```

Another nice effect would be the following order is the same for ```Put``` and ```Post```:
- collection.validate (afaik validation of isRequired field and such things)
- on validate event
- persist in db

Currently for ```Post``` it's:
- on validate event
- collection.validate
- persist in db

Any thoughts about this? I hope my description is understandable.

**Comment by andreialecu** [comment]
I'm not sure I understand. On Validate **is** called before On PUT and On POST. 

See the note [here](http://docs.deployd.com/docs/collections/adding-logic.html#s-On Validate -3789)

**Comment by Michael-Hamburger** [comment]
Yes, it is called. But it is called not in the same order. When using HTTP-Method POST on validate is called https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L642 and after that check the post() method is called.

When using PUT https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L637 the put() method is called and INSIDE this method the on validate event is called.

So when using HTTP-Method Post the order of processing is:
- on validate Event is called (when Validate-Script exists) (https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L642)
- method post() is called
- inside post() collection.validate is called (https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L599)

When using HTTP-Method Put the order of processing is:
- method put() is called
- inside put() collection.validate  is called (https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L552)
- on validate Event is called (when Validate-Script exists) (https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L587)
 
collection.validate should be the first and then the user written on validate event. I hope you can follow my description.

**Comment by andreialecu** [comment]
`collection.validate` is not what triggers the `On Validate` event. It does something else.

If you look at https://github.com/deployd/deployd/blob/master/lib/resources/collection/index.js#L586-L590 you will see that On Validate is called before On Put

**Comment by Michael-Hamburger** [comment]
Yes, I know it does not trigger the on validate event. It triggers some different validation. And that is the point. This validation should happen before the on validate Event. And that is not the case when using HTTP-Method post().

**Comment by andreialecu** [comment]
Ah I see what you mean now. Doesn't necessarily seem like a problem, but feel free to make a PR for this and I will review it.
